Realm/RLMSyncUser.h

```diff
@@ -124,21 +137,71 @@ NS_SWIFT_UNAVAILABLE("Use the full version of this API.");
  */
 - (NSArray<RLMSyncSession *> *)allSessions;
 
+// This set of permissions APIs uses immutable `RLMSyncPermissionValue` objects to
+// get and set permissions. It is intended to replace the set of APIs which directly
+// access Realms and Realm model objects to work with permissions.
+#pragma mark - Value-based Permissions API
+
+/**
+ Asynchronously retrieve all permissions associated with the user calling this method.
+
+ The results will be returned through the callback block, or an error if the operation fails.
+ The callback block will be run on the same thread the method was called on.
+
+ @warning This method must be called upon a thread with a currently active run loop. Unless
+          you have manually configured a run loop on a side thread, this will usually be the
+          main thread.
+
+ @see `RLMSyncPermissionResults`
+ */
+- (void)retrievePermissionsWithCallback:(RLMPermissionResultsBlock)callback;
+
+/**
+ Apply a given permission.
+
+ The operation will take place asynchronously, and the callback will be used to report whether
+ the permission change succeeded or failed. The user calling this method must have the right
+ to grant the given permission, or else the operation will fail.
+
+ @see `RLMSyncPermissionValue`
+ */
+- (void)applyPermission:(RLMSyncPermissionValue *)permission callback:(RLMPermissionStatusBlock)callback;
+
+/**
+ Revoke a given permission.
+
+ The operation will take place asynchronously, and the callback will be used to report whether
+ the permission change succeeded or failed. The user calling this method must have the right
+ to grant the given permission, or else the operation will fail.
+
+ @see `RLMSyncPermissionValue`
+ */
+- (void)revokePermission:(RLMSyncPermissionValue *)permission callback:(RLMPermissionStatusBlock)callback;
+
+// These permission APIs access Realms and Realm model objects representing
+// various permission states and actions, as well as standard Realm
+// affordances, to work with permissions. It is being deprecated in favor of
+// the `retrievePermissionsWithCallback:`, `applyPermission:callback:`, and
+// `revokePermission:callback:` APIs.
+#pragma mark - Realm Object-based Permissions API
+
 /**
  Returns an instance of the Management Realm owned by the user.
 
  This Realm can be used to control access permissions for Realms managed by the user.
  This includes granting other users access to Realms.
  */
-- (RLMRealm *)managementRealmWithError:(NSError **)error NS_REFINED_FOR_SWIFT;
+- (RLMRealm *)managementRealmWithError:(NSError **)error __deprecated_msg("Only use the management Realm for permission offers and responses.") NS_REFINED_FOR_SWIFT;
```

Comment:
I think we can use `NSClassFromString` in the Swift implementation to avoid the deprecation warnings and therefore drop the deprecation annotations in both the Objc & Swift versions of this method.